test/Base_Tests/src/Semantic/Multi_Value_As_Type_Refinement_Spec.enso

```diff
@@ -0,0 +1,337 @@
+id_a (x : A) -> A = x
+id_b (x : B) -> B = x
+
+type C
+    C_Ctor x
+
+    c_method self = "C method"
+
+C.from (that:B) = C.C_Ctor that
+
+add_specs suite_builder =
+    suite_builder.group "Multi Value as type refinement" group_builder->
+        group_builder.specify "conversion A -> B should not be available" <|
+            just_a = A.A_Ctor 1
+            Test.expect_panic Type_Error (just_a:B)
+            Test.expect_panic No_Such_Conversion (B.from just_a)
+            Test.expect_panic Type_Error (id_b just_a)
+
+        group_builder.specify "make_a_and_b->A&B presents as both A and B" <|
+            ab = make_a_and_b
+            ab.is_a A . should_be_true
+            ab.is_a B . should_be_true
+
+            ab.a_method . should_equal "A method"
+            ab.b_method . should_equal "B method"
+
+            (ab:A).to_text . should_equal "(A_Ctor 1)"
+            (ab:B).to_text . should_equal "(B_Ctor (A_Ctor 1))"
+            (id_a ab).to_text . should_equal "(A_Ctor 1)"
+            (id_b ab).to_text . should_equal "(B_Ctor (A_Ctor 1))"
+
+        group_builder.specify "after passing A&B value to a function expecting A argument, B becomes hidden" <|
+            ab = make_a_and_b
+            a2 = id_a ab
+            a2.is_a A . should_be_true
+            a2.is_a B . should_be_false
+
+            # Passing a2 to a function expecting B fails because B part was hidden
+            Test.expect_panic Type_Error (id_b a2)
+            Test.expect_panic No_Such_Method (a2.b_method)
+
+            # But it can be uncovered via explicit cast
+            (a2:B).b_method . should_equal "B method"
+            (id_b (a2:B)).to_text . should_equal "(B_Ctor (A_Ctor 1))"
+
+        group_builder.specify "after casting A&B to A, B part is again hidden" <|
+            ab = make_a_and_b
+            a2 = ab:A
+            a2.is_a A . should_be_true
+            a2.is_a B . should_be_false
+
+            # Passing a2 to a function expecting B fails because B part was hidden
+            Test.expect_panic Type_Error (id_b a2)
+
+            # But it can be uncovered via explicit cast
+            (a2:B).to_text . should_equal "(B_Ctor (A_Ctor 1))"
+            (id_b (a2:B)).to_text . should_equal "(B_Ctor (A_Ctor 1))"
+
+        group_builder.specify "passing A&B to function expecting B, then re-casting to A" <|
+            ab = make_a_and_b
+            b2 = id_b ab
+            a2 = b2:A
+            a2.is_a A . should_be_true
+            a2.is_a B . should_be_false
+
+            a2.a_method.should_equal "A method"
+            Test.expect_panic No_Such_Method (a2.b_method)
+
+        group_builder.specify "passing A&B to function expecting A, then re-casting to B" <|
+            ab = make_a_and_b
+            a2 = id_a ab
+            b2 = a2:B
+            b2.is_a A . should_be_false
+            b2.is_a B . should_be_true
+
+            Test.expect_panic No_Such_Method (b2.a_method)
+            b2.b_method.should_equal "B method"
```

Comment:
and an explicit cast back to A now fails?

(b2:A)

Reply:
```suggestion
            Test.expect_panic Type_Error (b2:A)
```

Reply:
No, actually the converse is true - we can still cast back to `A` because casting to `B` has only hidden the `A` part.

I've updated the test accordingly.